### PR TITLE
Remove hardcoded special distance metrics and check sklearn first

### PR DESCRIPTION
**Pull request by gokceneraslan**
If the number of datapoints are fewer than 4096, all pairwise distances are computed. sklearn-supported metrics are calculated via sklearn's `pairwise_distances` function and umap-only metrics are done via custom `dist.pairwise_special_metric` function. Instead of maintaining a list of umap-specific metrics here, I think it's easier to first check if sklearn supports a metric and then to fallback to custom function if not. Because, as we add new distance functions, it gets more cumbersome to maintain a list of special distance metrics that are not supported by sklearn e.g. poincare is missing right now.

There is a function in sklearn to get an incomplete list of supported metrics (https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/metrics/pairwise.py#L1147) but the full list is stored in a private list (https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/metrics/pairwise.py#L1237).

Therefore, I changed the code such that it first tries sklearn function first and falls back to custom function if it throws a ValueError (https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/metrics/pairwise.py#L1532).

### CODE REVIEW

**Comment by lmcinnes** [comment]
Good plan. Thanks for this. I believe the errors are my fault (pushing directly to dev without testing is, sadly, something I do all the time).
